RearrangingCars.java

```diff
@@ -55,6 +55,68 @@ public int hashCode() {
 }
 
 public class RearrangingCars {
+    public static ArrayList<Move> optimisedRearrange(int[] originalOrder, int[] finalOrder) {
+        if (originalOrder == null || finalOrder == null || originalOrder.length != finalOrder.length)
+            return null;
+
+        // Idea: Repeatedly replace current empty spot with the car that has to go there
+
+        ArrayList<Move> moves = new ArrayList<>();
+        if (originalOrder.length == 0) return moves;
+
+        HashMap<Integer, Integer> positions = new HashMap<>();
+        int N = originalOrder.length;
+        int[] originalOrderCopy = originalOrder.clone();
```

Comment:
I'm using a copy so as not to modify the input array, i think it's cleaner this way.